cdk/ec2/lib/compute-stack.ts

```diff
@@ -0,0 +1,143 @@
+import * as crypto from 'crypto';
+import * as cdk from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import {
+  Vpc,
+  InstanceType,
+  InstanceClass,
+  InstanceSize,
+  MachineImage,
+  SecurityGroup,
+  SubnetType,
+  Instance,
+  UserData,
+  LaunchTemplate
+} from 'aws-cdk-lib/aws-ec2';
+import { Role } from 'aws-cdk-lib/aws-iam';
+import { PrivateHostedZone, ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
+import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
+import { readFileSync } from 'fs';
+import { join } from 'path';
+import { AutoScalingGroup } from 'aws-cdk-lib/aws-autoscaling';
+
+interface ComputeStackProps extends cdk.StackProps {
+  vpc: Vpc;
+  ec2SecurityGroup: SecurityGroup;
+  ec2InstanceRole: Role;
+  hostedZone: PrivateHostedZone;
+  dbSecretArn: string;
+}
+
+export class ComputeStack extends cdk.Stack {
+  public frontendInstance: Instance;
+  public visitsAsg: AutoScalingGroup;
+  constructor(scope: Construct, id: string, props: ComputeStackProps) {
+    super(scope, id, props);
+
+    const { vpc, ec2SecurityGroup, ec2InstanceRole, hostedZone, dbSecretArn } = props;
+
+    // Import the database secret
+    const dbSecret = Secret.fromSecretCompleteArn(this, 'DBSecret', dbSecretArn);
+
+    // Grant the EC2 instance role permission to read the secret
+    dbSecret.grantRead(ec2InstanceRole);
+
+    // Define common properties for all instances
+    const instanceType = InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM);
+    const machineImage = MachineImage.latestAmazonLinux2023({});
+    const vpcSubnets = { subnetType: SubnetType.PRIVATE_WITH_NAT };
+
+    // List of services to deploy
+    const services = [
+      { name: 'setup', script: 'setup-user-data.sh', useAsg: false },
```

Comment:
We currently put admin server, config server, and discovery server on the same EC2 instance. I will still use the setup and rename it in cloudwatch agent config so that a meaningful name can show on service map.